### PR TITLE
Issue 33: process variables as out body

### PR DESCRIPTION
**Pull request by stephanpelikan**
see https://github.com/camunda/camunda-bpm-camel/issues/33

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -51,6 +51,7 @@ Parameter | Description
 `copyBodyAsVariable` | name of the process variable to which the body of the Camel should be copied. Default is `camelBody`
 `copyHeaders` | whether the [Camel message headers](http://camel.apache.org/header.html) should be copied as process variables
 `copyProperties` | whether the [Camel exchange](http://camel.apache.org/maven/current/camel-core/apidocs/org/apache/camel/Exchange.html) properties should be copied as process variables
+`copyVariablesToOutBody` | After the first wait state or the end of the process is reached the current values of the requested variables as a map is set to the out body instead of the process instance id. Supported formats are '*' (for all variables), 'var1,var2' (for certain variables) or 'var1' (for a single variable's value instead of a map). This requires the availability of Camunda's history service.
```
Comment (berndruecker) [inline README.md]: As you rely on the history service later on, this limitation should be mentioned here.

File: camunda-bpm-camel-common/src/main/java/org/camunda/bpm/camel/component/producer/StartProcessProducer.java [hunk 2]
```diff
@@ -76,7 +90,63 @@ public void process(Exchange exchange) throws Exception {
...
+          throw new RuntimeCamelException("Fetching process instance variables is not supported "
+                                          + "because a Camunda history service is needed but not available!");
+      }
+      
+      final String variableName = parameters.get(COPY_PROCESS_VARIABLES_TO_OUT_BODY_PARAMETER).toString();
+      
+      if (variableName.equals("*")) {
+        final List<HistoricVariableInstance> variables = historyService.createHistoricVariableInstanceQuery()
+                    .processInstanceIdIn(instance.getProcessInstanceId())
+                    .list();
```
Comment (berndruecker) [inline camunda-bpm-camel-common/src/main/java/org/camunda/bpm/camel/component/producer/StartProcessProducer.java]: When changing variables, this should result in the different versions of the variable instance - which is not what you want I think

File: camunda-bpm-camel-common/src/main/java/org/camunda/bpm/camel/component/producer/StartProcessProducer.java [hunk 3]
```diff
@@ -76,7 +90,63 @@ public void process(Exchange exchange) throws Exception {
...
     exchange.setProperty(EXCHANGE_HEADER_PROCESS_INSTANCE_ID, instance.getProcessInstanceId());
-    exchange.getOut().setBody(instance.getProcessInstanceId());
+    
+    setOutBody(exchange, instance);
+    
+  }
+  
+  private void setOutBody(final Exchange exchange, final ProcessInstance instance) {
+    if (parameters.containsKey(COPY_PROCESS_VARIABLES_TO_OUT_BODY_PARAMETER)) {
+      if (historyService == null) {
```
Comment (berndruecker) [inline camunda-bpm-camel-common/src/main/java/org/camunda/bpm/camel/component/producer/StartProcessProducer.java]: If the process instance is not yet ended but waits in a wait state you could also get this information from runtime. Just saying :-)

File: README.md (hunk 1 above)
Comment (stephanpelikan) [reply README.md]: Take a close look. I is already mentioned ;-)

File: camunda-bpm-camel-common/src/main/java/org/camunda/bpm/camel/component/producer/StartProcessProducer.java (hunk 3 above)
Comment (stephanpelikan) [reply camunda-bpm-camel-common/src/main/java/org/camunda/bpm/camel/component/producer/StartProcessProducer.java]: Yes, I know. It is also mentioned in the README.

File: camunda-bpm-camel-common/src/main/java/org/camunda/bpm/camel/component/producer/StartProcessProducer.java (hunk 2 above)
Comment (stephanpelikan) [reply camunda-bpm-camel-common/src/main/java/org/camunda/bpm/camel/component/producer/StartProcessProducer.java]: You are right. I'll check this.

File: README.md (hunk 1 above)
Comment (berndruecker) [reply README.md]: Pretty small :-) But you are right.

Comment (berndruecker) [reply README.md]: But I would rephrase it to something like "note that you need to set the history level (LINK) to at least XXX (todo. check which one is sufficient)

File: camunda-bpm-camel-common/src/main/java/org/camunda/bpm/camel/component/producer/StartProcessProducer.java (hunk 3 above)
Comment (stephanpelikan) [reply camunda-bpm-camel-common/src/main/java/org/camunda/bpm/camel/component/producer/StartProcessProducer.java]: I extended the implementation to support also runtime-only environments and accordingly adopted the README.

**Comment by berndruecker** [comment]
@stephanpelikan Hey Stephan, somehow this got stuck - not sure on which side :-( Shall we catch up or close this PR now?

File: camunda-bpm-camel-common/src/main/java/org/camunda/bpm/camel/component/producer/StartProcessProducer.java (hunk 2 above)
Comment (gwieshammer) [reply camunda-bpm-camel-common/src/main/java/org/camunda/bpm/camel/component/producer/StartProcessProducer.java]: In my understanding it returns only the last value of a variable, as it returns always at the first wait state of the process - isn't it?

**Comment by gwieshammer** [comment]
I would be happy if this PR could find its way to be merged...

**Comment by gwieshammer** [comment]
I could avoid the usage of historyService by changing from `startProcessInstanceByKey(...)` to `createProcessInstanceByKey(...).executeWithVariablesInReturn()`, which runs basically the same code but its signature returns a ProcessInstanceWithVariables. Along with the code and tests I also shortened the README.md to not mention the historyLevel any more for the new `copyVariablesToOutBody` property.

**Comment by berndruecker** [comment]
Hey @gwieshammer - great - thx! LGTM - and it aligns much better with how the platform works out of the box. 
The biggest drawback is maybe that we now use "executeWithVariablesInReturn" all the time, which is at least a change to the status quo. But I can#t think of any problem with it - so I am happy to merge this now.
